### ISSUE TITLE
SDR and TKE equations diverge for channel flow regression tests

### ISSUE THREAD

**Issue by sbidadi9** [Exawind/amr-wind#835]
The specific dissipation rate and turbulent kinetic energy diverge for the following regression tests:

_channel_kwsst
channel_kwsst_hypre
channel_kwsstiddes_

The initial residual for SDR increases rapidly within 10 iterations. The SDR equation diverges first followed by TKE. 

Below is the log file output after 10 iterations for _channel_kwsst_:

```
Step: 10 dt: 0.001133005691 Time: 0.006525006759 to 0.00765801245
CFL: 0.064903 (conv: 0.0648526 diff: 0 src: 0.00180802 )

Godunov:
  System                     Iters      Initial residual        Final residual
  ----------------------------------------------------------------------------
  MAC_projection               200        2.16265779e-05       3.440450749e-09
  tke_solve                      3            3928396270         0.01061103494
  sdr_solve                      4       1.974774149e+23            1201915264
  velocity_solve                 3         0.05594142048       3.450750796e-11
  Nodal_projection             200       3.846886758e-06       1.118572275e-09

Writing plot file       plt00010 at time 0.00765801245
WallClockTime: 10 Pre: 0.000185 Solve: 2.416 Post: 0.0145 Total: 2.43
Solve time per cell: 0.0002212
```

The CFL number is always less than 1.

**Comment by sbidadi9** [comment]
@marchdf Added more information on the log file output (above). The mac and nodal projection residuals reach max number of iterations after 8 iterations. Using hypre does not improve the solver performance.

**Comment by marchdf** [comment]
Sounds good I will take a look as well. Thanks!

**Comment by sbidadi9** [comment]
Thanks @marchdf

**Comment by mbkuhn** [comment]
This is the causal commit: https://github.com/Exawind/amr-wind/commit/43d13a306aff70d6bebaa7c7e991ec3d7dc67055
It was an important bug fix for the sdr source terms (so I'm not saying it is wrong), and I know that it fixed other things, but this change definitely led to the divergence of sdr that we see in the channel_kwsst family of tests (channel_kwsst, channel_kwsst_hypre, channel_kwsstiddes). Note: these are the only reg tests that use the sdr_solve, and every single one diverges in the sdr solve for at least one time step within the first 10 time steps. The divergence of the sdr leads to the MAC and nodal projections maxing out their iterations.

**Comment by marchdf** [comment]
I've been playing around with this and have an exact reproducer in nalu-wind. That one seems to have no problems. If I coarsen the y direction (thereby reducing the aspect ratios), it seems to run fine... Is that the issue? Maybe we can't handle those aspect ratios? I still need to do a bunch of due diligence...

**Comment by ashesh2512** [comment]
Two pull requests are open at this point - #842 and #852. Both touch similar topics.

**Comment by sbidadi9** [comment]
The PR #852 fixes the issue.

**Comment by wenchaoyu** [comment]
Hi @sbidadi9 and @marchdf,
It seems with the recent version of amrwind, these tests _channel_kwsst_ and _channel_kwsst_hypre_ still diverge after some iterations. Taking _channel_kwsst_ for example, MAC_projection reaches the max number after 25 iterations and simulation crashes at 44 iterations: 
![image](https://github.com/Exawind/amr-wind/assets/4232919/b84bb41b-ca8c-447b-b8cc-3de592bd658c)

If changing the diff_type from implicit to CN, and using _hypre_, the simulation can handle longer but still crashes after 937 iterations:
![image](https://github.com/Exawind/amr-wind/assets/4232919/7f55be29-3a19-4cc3-8561-0eaaa41c94cf)

I tested also the channel case in nalu-wind https://github.com/Exawind/nalu-wind/tree/master/reg_tests/test_files/SSTChannelEdge, the simulation crashes after hundereds of iterations. 

Do you have this issue on your side ?
Thanks

**Comment by sbidadi9** [comment]
Hi @wenchaoyu 

I ran the _channel_kwsst_hypre_ just now for 100 iterations with implicit time stepping, and the cash does not crash. Below is the output.
![image](https://github.com/Exawind/amr-wind/assets/96149491/3d892937-9bb9-45a8-9b23-2b1b71a89917)

Your MAC projection residuals are maxing out. I would suggest running it with hypre.

Please reach out to Marc about the Nalu-Wind code.

**Comment by wenchaoyu** [comment]
Thanks @sbidadi9 for your reply.

I have tried also this case, I got similar output as you at 100 iteration.
However, Nodal projection will max out if running slightly longer the simulation, at 164 iteration, below is the output. 
![image](https://github.com/Exawind/amr-wind/assets/4232919/ac9cf681-061a-4a58-b0da-6b10ff94f484)

and it will crash later, sdr diverges
![image](https://github.com/Exawind/amr-wind/assets/4232919/0958e3af-5044-48a2-8c89-a598a89f9ac8)

It seems with hypre and implicit time stepping, it helps to delay the divergence, but cannot avoid it.
